internal/controller/dashboard_controller.go

```diff
@@ -147,162 +145,41 @@ func (r *DashboardReconciler) SetupWithManager(mgr ctrl.Manager) error {
-		return labels[grafanaOrganizationLabel], nil
-	}
-
-	// Return an error if no label was found
-	return "", errors.New("No organization label found in configmap")
-}
-
-func (r DashboardReconciler) configureDashboard(ctx context.Context, grafanaAPI *grafanaAPI.GrafanaHTTPAPI, dashboardCM *v1.ConfigMap) error {
-	logger := log.FromContext(ctx)
-
-	dashboardOrg, err := getOrgFromDashboardConfigmap(dashboardCM)
-	if err != nil {
-		logger.Error(err, "Skipping dashboard, no organization found")
-		return nil
+		return errors.WithStack(err)
 	}
 
-	// TODO Tenant Governance: Filter the dashboards with the list of authorized tenants
+	service := grafana.NewService(r.Client, grafanaAPI)
 
-	// Switch context to the dashboards-defined org
-	organization, err := grafana.FindOrgByName(grafanaAPI, dashboardOrg)
+	// Configure the dashboard in Grafana
+	err = service.ConfigureDashboard(ctx, dashboard)
 	if err != nil {
-		logger.Error(err, "failed to find organization", "organization", dashboardOrg)
 		return errors.WithStack(err)
 	}
-	currentOrgID := grafanaAPI.OrgID()
-	grafanaAPI.WithOrgID(organization.ID)
-	defer grafanaAPI.WithOrgID(currentOrgID)
-
-	for _, dashboardString := range dashboardCM.Data {
-		var dashboard map[string]any
-		err = json.Unmarshal([]byte(dashboardString), &dashboard)
-		if err != nil {
-			logger.Error(err, "Failed converting dashboard to json")
-			continue
-		}
-
-		dashboardUID, err := getDashboardUID(dashboard)
-		if err != nil {
-			logger.Error(err, "Skipping dashboard, no UID found")
-			continue
-		}
-
-		// Clean the dashboard ID to avoid conflicts
-		cleanDashboardID(dashboard)
-
-		// Create or update dashboard
-		err = grafana.PublishDashboard(grafanaAPI, dashboard)
-		if err != nil {
-			logger.Error(err, "Failed updating dashboard")
-			continue
-		}
-
-		logger.Info("updated dashboard", "Dashboard UID", dashboardUID, "Dashboard Org", dashboardOrg)
-	}
 
 	return nil
 }
 
 // reconcileDelete deletes the grafana dashboard.
-func (r DashboardReconciler) reconcileDelete(ctx context.Context, grafanaAPI *grafanaAPI.GrafanaHTTPAPI, dashboardCM *v1.ConfigMap) error {
-	logger := log.FromContext(ctx)
-
+func (r DashboardReconciler) reconcileDelete(ctx context.Context, grafanaAPI *grafanaAPI.GrafanaHTTPAPI, dashboard *v1.ConfigMap) error {
 	// We do not need to delete anything if there is no finalizer on the grafana dashboard
-	if !controllerutil.ContainsFinalizer(dashboardCM, DashboardFinalizer) {
+	if !controllerutil.ContainsFinalizer(dashboard, DashboardFinalizer) {
 		return nil
 	}
 
-	dashboardOrg, err := getOrgFromDashboardConfigmap(dashboardCM)
-	if err != nil {
-		logger.Error(err, "Skipping dashboard, no organization found")
-		return nil
-	}
+	service := grafana.NewService(r.Client, grafanaAPI)
 
-	// Switch context to the dashboards-defined org
-	organization, err := grafana.FindOrgByName(grafanaAPI, dashboardOrg)
+	// Unconfigure the dashboard in Grafana
+	err := service.UnconfigureDashboard(ctx, dashboard)
```

Comment:
You mean DeleteDashboard right?

Reply:
Yes that sounds better, I just made up this name (which I knew was very clunky)